File: python/modelos.py
from abc import ABC, abstractmethod

class ProgramaAcademico(ABC):
    def __init__(self, nombre: str):
        self._nombre = nombre

    @property
    def nombre(self):
        return self._nombre

    @abstractmethod
    def evaluar_aprobacion(self, notas: list) -> bool:
        """Metodo polimorfico que cada subclase implementara."""
        pass


class Curso(ProgramaAcademico):
    def __init__(self):
        super().__init__("Curso")

    def evaluar_aprobacion(self, notas: list) -> bool:
        if not notas or len(notas) < 3:
            return False
        promedio = sum(notas) / len(notas)
        return promedio >= 10.0  


class Diplomado(ProgramaAcademico):
    def __init__(self):
        super().__init__("Diplomado")

    def evaluar_aprobacion(self, notas: list) -> bool:
        if not notas or len(notas) < 3:
            return False
        promedio = sum(notas) / len(notas)
        return promedio >= 14.0  


class Bootcamp(ProgramaAcademico):
    def __init__(self):
        super().__init__("Bootcamp")

    def evaluar_aprobacion(self, notas: list) -> bool:
        if not notas or len(notas) < 3:
            return False
        
        return all(nota >= 14.0 for nota in notas)


class Persona:
    def __init__(self, cedula: str, nombre_completo: str, correo: str):
        self._cedula = cedula
        self._nombre_completo = nombre_completo
        self._correo = correo

    @property
    def cedula(self): return self._cedula

    @property
    def nombre_completo(self): return self._nombre_completo

    @property
    def correo(self): return self._correo


class Alumno(Persona):
    def __init__(self, cedula: str, nombre_completo: str, correo: str, programa: ProgramaAcademico):
        super().__init__(cedula, nombre_completo, correo)
        self.programa = programa
        self.notas = []  

    def agregar_nota(self, nota: float):
        if len(self.notas) < 3:
            self.notas.append(nota)
            return True
        return False

    def calcular_promedio(self) -> float:
        if not self.notas:
            return 0.0
        return sum(self.notas) / len(self.notas)

    def esta_aprobado(self) -> bool:
        return self.programa.evaluar_aprobacion(self.notas)


class Profesor(Persona):
    def __init__(self, cedula: str, nombre_completo: str, correo: str, especialidad: str, materia: str):
        super().__init__(cedula, nombre_completo, correo)
        self.especialidad = especialidad
        self.materia = materia

File: java/modelos.java
import java.util.ArrayList;
import java.util.List;

abstract class Persona {
    protected String cedula;
    protected String nombreCompleto;
    protected String correo;

    public Persona(String cedula, String nombreCompleto, String correo) {
        this.cedula = cedula;
        this.nombreCompleto = nombreCompleto;
        this.correo = correo;
    }

    public String getCedula() { return cedula; }
    public String getNombreCompleto() { return nombreCompleto; }
    public String getCorreo() { return correo; }
}

class Profesor extends Persona {
    private String especialidad;
    private String materia;

    public Profesor(String cedula, String nombreCompleto, String correo, String especialidad, String materia) {
        super(cedula, nombreCompleto, correo);
        this.especialidad = especialidad;
        this.materia = materia;
    }

    public String getEspecialidad() { return especialidad; }
    public String getMateria() { return materia; }
}

abstract class ProgramaAcademico {
    protected String nombre;

    public ProgramaAcademico(String nombre) {
        this.nombre = nombre;
    }

    public String getNombre() { return nombre; }
    public abstract boolean evaluarAprobacion(List<Double> notas, double promedio);
}

class Curso extends ProgramaAcademico {
    public Curso() { super("Curso"); }

    @Override
    public boolean evaluarAprobacion(List<Double> notas, double promedio) {
        return promedio >= 10.0;
    }
}

class Diplomado extends ProgramaAcademico {
    public Diplomado() { super("Diplomado"); }

    @Override
    public boolean evaluarAprobacion(List<Double> notas, double promedio) {
        return promedio >= 14.0;
    }
}

class Bootcamp extends ProgramaAcademico {
    public Bootcamp() { super("Bootcamp"); }

    @Override
    public boolean evaluarAprobacion(List<Double> notas, double promedio) {
        if (promedio < 14.0) return false;
        for (double n : notas) {
            if (n < 14.0) return false;
        }
        return true;
    }
}

class Alumno extends Persona {
    private ProgramaAcademico programa;
    private List<Double> notas;

    public Alumno(String cedula, String nombreCompleto, String correo, ProgramaAcademico programa) {
        super(cedula, nombreCompleto, correo);
        this.programa = programa;
        this.notas = new ArrayList<>();
    }

    public void agregarNota(double nota) {
        if (this.notas.size() < 3) {
            this.notas.add(nota);
        }
    }

    public double calcularPromedio() {
        if (notas.isEmpty()) return 0.0;
        double suma = 0.0;
        for (double n : notas) suma += n;
        return suma / notas.size();
    }

    public boolean estaAprobado() {
        if (notas.size() < 3) return false;
        return programa.evaluarAprobacion(notas, calcularPromedio());
    }

    public ProgramaAcademico getPrograma() { return programa; }
    public List<Double> getNotas() { return notas; }
}